exercises/practice/bowling/src/lib.rs

```diff
@@ -4,18 +4,76 @@ pub enum Error {
     GameComplete,
 }
 
-pub struct BowlingGame {}
+const FULL_PINS: u16 = 10;
+const NUMBER_OF_FRAMES: usize = 10;
+const TWO_THROWS: usize = 2;
+
+pub struct BowlingGame {
+    throws: Vec<u16>,
+    first_throw: bool,
+}
 
 impl BowlingGame {
     pub fn new() -> Self {
-        unimplemented!();
+        BowlingGame {
+            throws: Vec::new(),
+            first_throw: true,
+        }
     }
 
     pub fn roll(&mut self, pins: u16) -> Result<(), Error> {
-        unimplemented!("Record that {} pins have been scored", pins);
+        if pins > FULL_PINS
+            || (!self.first_throw && (self.throws.last().unwrap() + pins) > FULL_PINS)
+        {
+            return Err(Error::NotEnoughPinsLeft);
+        }
+        if self.score().is_some() {
+            return Err(Error::GameComplete);
+        }
+        self.throws.push(pins);
+        if pins == FULL_PINS {
+            self.first_throw = true;
+        } else {
+            self.first_throw = !self.first_throw;
+        }
+
+        Ok(())
     }
 
     pub fn score(&self) -> Option<u16> {
-        unimplemented!("Return the score if the game is complete, or None if not.");
+        let mut score = 0;
+        let mut throws_index = 0;
+        // repeat NUMBER_OF_FRAMES times
+        for _frame in 0..NUMBER_OF_FRAMES {
+            match (
```

Comment:
That's shouldn't be a match. Store each throw in a variable instead and then check the validity.